### ISSUE TITLE
Add support for supplemental CSS selectors

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#4782]
What about using special body classes to include new CSS? If new sites got the classes, but existing sites did not, we could prevent the CSS from affecting existing sites. (backwards-compatible)

At install time, the list of special body classes would be determined. Then, for the rest of that site's lifetime, only those original classes would be applied. 

If new CSS is added in 1.17, one of those classes could be `version-1-17`. All the new CSS rules in 1.17 could be prefixed with `.version-1-17 `, and all sites installed at version 1.17 or later would have this class applied. However, if I had installed Backdrop at version 1.14, my site would not have the `version-1-17` body class, so would not get the CSS rules applied. 

Backdrop sites installed at version 1.18 would also get all previous body classes, and if there were new CSS changes introduced in 1.18 too, that site would also get a `version-1-18` class (but if there weren't any CSS changes, no class would be added). 

**Pro**: This approach would allow us to keep all the related CSS in the same file, rather than moving the newer ones into a separate file that would not be included on newer sites.

**Pro**: This approach could work for CSS included in any part of core (modules, themes, etc).

**Con**: This approach could eventually lead to a lot of body classes on older Backdrop sites. 

Related issues:

- [Decision](https://github.com/backdrop/backdrop-issues/issues/4167)
- [Supplemental style sheets](https://github.com/backdrop/backdrop-issues/issues/4512)

**Comment by BWPanda** [comment]
I like this idea, especially the two pros.

As for the con, we could try shorter classes (like `v117`), or we could make it so that only the current/latest version is added as a class on install. However that would mean needing to update all existing version-specific CSS to add the new version class, so maybe not a good option...

**Comment by jenlampton** [comment]
> As for the con, we could try shorter classes (like `v117`)

Yeah, I like that better.

> However that would mean needing to update all existing version-specific CSS to add the new version class on each new release, so maybe not a good option...

I thought about that too. I think I prefer less work with each release. Updating all existing selectors seems like it would be easy to forget.

**Comment by BWPanda** [comment]
> Updating all existing selectors seems like it would be easy to forget.

Indeed. And they'd be spread all over the place.

**Comment by klonos** [comment]
I also like this idea 👍 ...I would love it if we figured a way to not "punish" new sites with adding all the version classes though; wondering if we could reverse this, so that only old sites that choose to not update have the extraneous classes. Needs a bit more thought 🤔

**Comment by stpaultim** [comment]
This proposal (above) is interesting. Part of me wonders if it is really better than supplemental stylesheets, but I can see why folks might think so.  

In the spirit of brainstorming, I'll throw out a few potentially bad ideas (modifications to this proposal) that might not be bad or might spark some good ideas. 

1) Add a yearly class instead of a version based class. Simply to make it very clear that these kinds of changes only get in once a year and to limit the amount of extra classes that we might eventually accumulate. 

2) I understand the proposal to suggest adding a new class to the body field for each version (or year), meaning that in 2018 (if we are still on 1.x) we might have sites with `<body class="v2021 v2022 v2023 v2024 v2025 v2026 v2027 v2028>`. We could (I think) shift the focus to the CSS files and only add a single class to each site, depending upon when it was installed. Then we could multiply the classes in the CSS, so that the changed rules simply apply to more different classes.

This option would keep the page markup cleaner, but would create a greater amount of 'redundant' css. Maybe this is better, because the markup is more likely to be seen/noticed. Again, this is probably a bad idea - but, maybe someone can make it better?

Markup contains only one year based class (example = `<body class="2022">`).

```

.v2021 h1, .v2022 h1, .v2023 h1 {
	padding: red;
}

.v2021 .container .block p, .v2022 .container .block p, .v2023 .container .block p {
	padding: blue;
}

.v2022 header, .v2023 header {
	padding: green;
}

.v2023 .menu-item {
	padding: yellow;
}
```

- - - - - - - - - - - - - -

> figured a way to not "punish" new sites with adding all the version classes though; wondering if we could reverse this, so that only old sites that choose to not update have the extraneous classes

I don't know if I understand the logic here. These new classes/changes potentially break old sites. I am not sure that we would give the old site the opportunity to choose the new classes (we could, but is the complexity worth giving them a fix to a problem that they have already solved). So, I don't think that there is necessarily any choice involved in this. 

Also, why would you "punish" old sites IF they did choose to not risk breaking their site over improvements which they most likely don't need, because they have already solved the problem.

**Comment by BWPanda** [comment]
In order to reduce the number of classes added to sites, what if we only supported backwards-compatibility for the previous (e.g. 2) versions...?

So each (minor) release of Backdrop, we'd go through and remove any CSS class prefixes that were older than 2 version ago. This would mean that those changes were optional when they were first introduced, and for the next 2 versions of Backdrop (giving people a chance to update their themes, etc.), but then they're added to core after that.

For example:
- 26th Feb 2021: We add CSS that fixes something, but we give it the `1.19.0` class prefix so only new sites (as of v1.19.0) get that fix
- 15th Jan 2022: We're about to release v1.21.0, so we go through and remove any `1.19.0` classes from CSS files. Those CSS changes are now applied to all sites (as of v1.21.0)

This means we should only ever have 2 supplemental classes on the body tag. We can change this to support more versions of Backdrop (maybe 6, so then we're supporting the last 2 years of releases), but at least there's a limit to the number of CSS classes we're adding, and the amount of time CSS changes are optional before being made mandatory.

**Comment by jenlampton** [comment]
> In order to reduce the number of classes added to sites, what if we only supported backwards-compatibility for the previous (e.g. 2) versions...?

I think you mean only supporting two minor versions -- and if that's the case I am very much against this idea. All of my clients really dislike surprise costs, and they like surprise breakage even more. 

Once they get their website working, they are happy to pay for routine updates, but the expectation is that the updates will fix bugs, not cause more. When updates start causing bugs is when people stop being willing to update anything. :(

**Comment by jenlampton** [comment]
@stpaultim  there is a PR/sandbox ready for testing: https://github.com/backdrop/backdrop/pull/4887

What it does:
* adds a new config value for `initial_version`
   - New installs get the Backdrop version at install time
   - The update hook sets this to 1.29.0 for existing sites
* Reports this version via telemetry
* Basis uses this version to set a body class
* Basis uses the new body class to conditionally alter the display of breadcrumbs.

**Comment by klonos** [comment]
Wondering whether we should consider using `data-*` attributes instead of classes here? 🤔 ...something like `data-initial-version="1.30.0"` for example? Then we could do things like the following:
```css
[data-initial-version]:not([data-initial-version="1.30.0"]) ul.primary li a {
 /* CSS rules here */
}
[data-initial-version="1.30.0"] ul.primary li a {
 /* CSS rules here */
}
```

Classes look simpler/prettier, but `data-*` attributes have been specifically designed to hold data values (like the version number in our use case here).

**Comment by jenlampton** [comment]
Interesting. I like the data attribute because the value fits better. But we're adding CSS here, and that's what classes are for. I'd prefer we stick to javascript being added via data attributes.

**Comment by stpaultim** [comment]
I forgot about this PR. Will try to test soon. @jenlampton

**Comment by stpaultim** [comment]
@jenlampton Very excited about this PR. I finally got around to testing.

Just a note that this PR does not work in sandbox or in a site set up with the dev branch of BackdropCMS, because the initial_version is set to 1.30.x-dev. 

![image](https://github.com/user-attachments/assets/09a1a5b6-b3b8-4cdb-b794-a86196102d0f)

If I change this to: "1.30.0", then the class is activated. It seems as if the class inserted is always equal to the version number. 

V. 1.30.3 inserts class "installed-1-30-2"
V. 1.31.0 inserts class "installed-1-31-0"
V. 1.45.65 inserts class "installed-1-45-65"

Unless I am misunderstanding something, this is not good, because CSS changes are based off a specific class. Changes that are targeted at the class 1.30.0 will no longer work on sites installed with version 1.30.1 if the site is now using a class of `installed-1-30-2`.

I believe that idea would be that once a year, we create a new class. Let's say `supplemental-1-30` and any site installed with version 1.30.0 or later would get that class. 

One year later we would add class supplemental-1-33 and any site installed with version of 1.33.0 or later would get two classes `supplemental-1-30` and `supplemental-1-33`.

One year later we would add class supplemental-1-36 and any site installed with version of 1.36.0 or later would get two classes `supplemental-1-30` and `supplemental-1-33` and `supplemental-1-36`.  

Changing the class for every release would be good if we could write css that affects class = `installed-1-30.0 or greater`. But, I don't think that is possible. 

See css included in this PR:

```
.installed-1-30 ul.primary li a,
.installed-1-30 ul.primary li a:visited,
.installed-1-30 ul.primary li a:hover,
.installed-1-30 ul.primary li.active a,
.installed-1-30 .expand-dropdown-tabs-control,
.installed-1-30 .responsive-tabs-mobile .expand-dropdown-tabs-control:before,
.installed-1-30 .responsive-tabs-mobile .expand-dropdown-tabs-label {
  background-color: #c5c5d0;
}
.installed-1-30 ul.primary li a:hover {
  background-color: #a7a7b0;
}
```

With the current PR, I believe that this would break once version 1.30.1 comes out.

**Comment by quicksketch** [comment]
> I believe that idea would be that once a year, we create a new class. Let's say `supplemental-1-30` and any site installed with version 1.30.0 or later would get that class.

I would be okay allowing new classes to be added on an as-needed basis. e.g. we might have supplemental classes for `1-30`, `1-31`, and `1-32` (if we fix a bunch of stuff quickly), but then we might slow down and only add `1-35`, `1-40`... adding more classes whenever we introduce a new CSS fix. I don't think we need to fix it to a yearly cycle, which might discourage contribution and also be difficult to schedule the fixes being included.

**Comment by jenlampton** [comment]
Since @quicksketch likes the idea, I turned the PR from a loose proof-of-concept to something more suitable for core :tada

* I went ahead and fixed the update hook @klonos -- [see the new changes here](https://github.com/backdrop/backdrop/pull/4887/files#diff-71ecd72989e99f3bd8c92afc37144a39d6c51af1463df2b5283548dc06467830R3638).
* I changed the preprocess to loop over all the versions where we add changes @stpaultim -- [see the new changes here](https://github.com/backdrop/backdrop/pull/4887/files#diff-e39cacd434aa70a006cdc5ece9044bb339ea0457f1f4108ed825234689cc08c1R41)

Feedback requested on: 
* I named the classes `update-1-30` etc, but that's open for discussion.
* I named the system core variable `initial_version` but that could be `install_version` (to match `install_profile`) if people prefer.

**Comment by stpaultim** [comment]
@quicksketch and @jenlampton Thanks. So nice to see this moving forward.

I would agree, that if there is no objection, we don't need to tie new css to once per year, but could make the decision based upon need. The decision to include css updates could be based upon either volume of PR's ready to go OR urgency of change. But, that isn't so important right now. 

If I understand correctly, this work technically by:

1) Keeping an array of all version numbers to which css changes have been added. 
2) Keeping track of the initial version number, when the site was installed.
3) Adding classes for each version that had css changes AFTER the installed version. 

That pretty much sums it up, I think. Do I have this correct?

I don't feel strong what the classes look like. I recognize that we're probably trying to make the classes semantic. BUT, I think compactness is also nice for aesthetic reasons in the code. I'd be ok with something much shorter that is less understandable, but explained in comments. 

`cssv130` `cssv131` `cssv135`
Or
`cssv-1-30` `cssv-1-31` `cssv-1-35`
Or
`css-1-30` `css-1-31` `css-1-35`

CURRENT PROPOSAL:

`update-1-30` `update-1-31` `update-1-35`

Again, I do NOT feel strongly about any of these. But, just want to make sure we've considered options before settling on any one choice.

When looking at them next to one another, I'm not sure that the compact versions are compact enough to matter. So, I'm happy with any of these, including the current proposal.

**Comment by jenlampton** [comment]
> Adding classes for each version that had css changes AFTER the installed version.

This is the opposite. We want only the changes that come BEFORE the site was installed, because anything that came after has a chance of breaking the site. 

> PR still does not work on dev release, because of "initial_version": "1.30.x-dev" . Do we care?

That's strange PHP's `version_compare()` is supposed to account for `-dev` versions.  I'll have another look at that. 

> Is there anything else to test, other than the creation and insertion of the class?

We definitely need more testing around circumstances when the classes get added. For testing, maybe we need a separate PR that adds a class for 1.28 and 1.29? We should test installs at 1.27, 1.28, 1.29 and 1.30, and then also updating and confirming on the appropriate classes are added. 

We should also test:
* The telemetry data
* The update hook that sets the installed version (for older sites) since we are going to be sending that to telemetry

**Comment by stpaultim** [comment]
At last weeks DEV meeting @quicksketch mentioned some interest in moving this issue along, but mentioned that some help might be needed. 

See the discussion at the last meeting here:
https://www.youtube.com/live/VRHIfFzfhn0?si=4zvSj5cyCKfA0P5b&t=2101

**Comment by stpaultim** [comment]
There is a lot of interest in getting this into 1.30.0, but it's not clear that either @jenlampton or @quicksketch have time to work on it. I'm not sure how this PR fits with my skill set, but will take a look. To start with, I'm generating a list of known issues or suggested changes for either myself or someone else to work through. 

Issues that need to be addressed:

* Current PR does not seem to be working in sandbox sites, but does work on local sites. https://github.com/backdrop/backdrop-issues/issues/4782#issuecomment-2484641807
* Decide on name for setting: https://github.com/backdrop/backdrop/pull/4887#pullrequestreview-2444597502
* Small formatting changes: https://github.com/backdrop/backdrop/pull/4887#discussion_r1849892718
* Decision about "install time": https://github.com/backdrop/backdrop/pull/4887#pullrequestreview-2485011646
* Typo: https://github.com/backdrop/backdrop/pull/4887#pullrequestreview-2485021198
* Better names or descriptions of variables: https://github.com/backdrop/backdrop/pull/4887#pullrequestreview-2485068988 
* Data attributes? - https://github.com/backdrop/backdrop-issues/issues/4782#issuecomment-2421640191

Specific things that need to be tested. https://github.com/backdrop/backdrop-issues/issues/4782#issuecomment-2486539922
* We definitely need more testing around circumstances when the classes get added.
  * The telemetry data
  * The update hook that sets the installed version (for older sites) since we are going to be sending that to telemetry
  
Decision needs to be made on format for CSS Selectors. Which do we want to use:
* update-1-30
* supplemental-1-30
* cssv-1-30
* cssv130
* Other...

@quicksketch mentioned a preference for `supplemental-1-30`, but this still seems to be open for discussion.

----

Having created this list, I'm thinking that this PR may fall within my skillset. I may need help when it comes to tests OR maybe sooner. I'll try to do something tonight.

**Comment by herbdool** [comment]
What if the initial install variable was part of the theme config instead of the system config? Maybe there are some downsides but I was thinking that someone may want to create a new theme on an old install and would like to benefit from all the fixes in Basis or otherwise.

**Comment by stpaultim** [comment]
> What if the initial install variable was part of the theme config instead of the system config? Maybe there are some downsides but I was thinking that someone may want to create a new theme on an old install and would like to benefit from all the fixes in Basis or otherwise.

@herbdool I'm interested, but not fully understanding how making the config variable part of the theme config helps in this case. I'm not sure if it's in the existing PR yet, but I believe that we have in mind a configuration option that would allow one to add all the changes, even if they have an old install. 

If we allow anyone to use a configuration setting to accept all changes to basis, would that accomplish your goal?

When you suggest putting this variable in the theme config, are you suggesting in the config file for Basis?

**Comment by herbdool** [comment]
> If we allow anyone to use a configuration setting to accept all changes to basis, would that accomplish your goal?

@stpaultim It doesn't matter to me where the variable is, so long as someone can have a new install version so they can accept all changes up to that point. What you're saying seems to be slightly different.

**Comment by stpaultim** [comment]
@herbdool Thanks. The way the PR is supposed to work (you may understand this already, but I want to clarify to be sure everyone is on the same page), is that if your original install was at version 1.30, you will automatically get all changes to basis that have been implemented in 1.30 or earlier. In this case, there will not be any earlier changes. 

If your initial installation is 1.32, you will get all changes made in versions 1.32 or earlier. So by default, newer installations are getting all changes that have been made up to the point of their installation, but will not get changes (by default) made AFTER they installed. 

Someone who installed at version 1.24 will not get any changes, by default, because they installed before any changes were accepted. BUT, I think they should have the option of enabling all changes that came in later version. I'm not sure if that option exists yet in the PR, but I would like to see it. Either now or as a short term follow-up issue.

**Comment by herbdool** [comment]
I wasn't thinking of an option to "accept all later changes", just the ability to change the initial version. Which one can do by manually changing the number in the config file. So if the initial install is 1.30 but a while later, someone wants to start with a fresh theme on that site, they could change `initial_version` to the version of the codebase at that point in time, say 1.34. Then it would have all changes up to 1.34 but no later.

**Comment by stpaultim** [comment]
> So if the initial install is 1.30 but a while later, someone wants to start with a fresh theme on that site, they could change initial_version to the version of the codebase at that point in time, say 1.34. Then it would have all changes up to 1.34 but no later.

I see your original point now. I was going to suggest that user would be able to do this anyway. Then I realized, that if they do and another module uses this variable for something else, it could cause unexpected problems. Now, I think I understand, that is why you suggested making this setting specific to theme. 

Let's get the next version of this PR working and then see where we are at on this issue. It may be worth taking about during next dev meeting.

**Comment by herbdool** [comment]
I played around with putting the variable in the theme settings. So if it's done this way, each child theme of Basis and Basis itself would get the initial version saved to it's settings (in the system update hook), if the theme is enabled.

This way, multiple themes can exist side by side and each can specify what css updates to include. Depending on when they were created.

**Comment by quicksketch** [comment]
There's some other benefit to globally storing the install version: we could track it in telemetry. However, if users are manipulating that value directly to adjust their themes, that would eliminate the whole point of tracking it.

I think there needs to be a theme setting separate from the initial version config value. If that's exposed in the UI, the theme should only show versions where CSS changes were actually made. There should also be a way to just say "always include latest CSS", which would be useful for the majority of users of Basis (those not sub-theming it). Ideally contrib themes that subtheme Basis would would update that theme setting to match the version they are compatible with, though I don't think these have that ability to execute PHP when enabled.

**Comment by stpaultim** [comment]
@jenlampton or @quicksketch Can either or both of you look at my list of things to do here: https://github.com/backdrop/backdrop-issues/issues/4782#issuecomment-2540210988

Were there things that are not yet in the PR that need to be?

I have refreshed the PR. Now checking to make sure I didn't break it. 

I did the following:

- Changed `$initial_version` to `$install_version` as suggested by @quicksketch 
- Changed system update checks to establish install versions for old sites to reflect suggestion by @quicksketch 
- Changed `updated_versions` to `supplemental_css_versions` as suggested by @indigoxela 
- Fixed type found by @indigoxela 
 
Now trying to figure out next steps.

**Comment by stpaultim** [comment]
@quicksketch - Would we ever be including css changes in bugfix releases OR will this be restricted only to minor releases?

Will we ever have a class that looks like `update-1.30.3` or would it stop at `update-1.30` (Just using `update` as a working example for now)?

**Comment by quicksketch** [comment]
> Would we ever be including css changes in bugfix releases OR will this be restricted only to minor releases?

I expect it depends on the frequency of the changes. But I don't see any reason inherently to restrict it to minor releases. Once we start tracking install version (accurately) on new installs, we can add classes for any specific version.

**Comment by stpaultim** [comment]
@quicksketch and others. Are we looking at a setting like this for Basis. I am working on this in PR, but have not yet pushed this to sandbox.

![image](https://github.com/user-attachments/assets/c2f41abb-9f13-4230-afce-d629bd87ba5e)

![image](https://github.com/user-attachments/assets/a01f0489-5f20-4192-84ec-8d4eac6a9c88)

I would appreciate any suggestions on the UI for this now. I will be working on this off and on over the next few hours.

**Comment by stpaultim** [comment]
Another more complicated version:

Version options only visible if "Specific version" is selected.

![image](https://github.com/user-attachments/assets/554b9cf0-706f-44fb-a392-563b0caf75ec)

Still work in progress. Posting half baked ideas in case anyone is available today to offer immediate feedback and keep this PR moving.

**Comment by herbdool** [comment]
@stpaultim that's along the lines of what I was imagining, though we might want to have them visible but disabled and have a separate checkbox that says something like "Expert option: change the core version that is being tracked". (Or worded better). Or is that too much? 

I believe a default setting can be made in a theme info file so one of @stpaultim's contrib themes could specify the version that it will work with. That's only if it's a child theme of Basis or if it wants to do something similar to Basis.

**Comment by stpaultim** [comment]
@herbdool How about something like this. Closed by default. Includes a link to a page in the documentation that provide more details (page not yet written). 

![image](https://github.com/user-attachments/assets/03ccbbc4-4b8d-4b1c-a594-f10a2d9db591)

**Comment by quicksketch** [comment]
I like the description text. I like the version dropdown showing only certain versions.

I think the "Thru installation" option is a bit too confusing. I would be tempted to take that option out and use the "Specific version" option set to the installation version. Maybe just put in the description text: "Defaults to [version]."

And I think having an "All changes" or "Latest" version option is desirable, for sites that use an unmodified Basis.

**Comment by herbdool** [comment]
@stpaultim I've posted comments in the PR as well.

Screenshot looks better. I agree with Nate's comments.

**Comment by stpaultim** [comment]
OK, I have a working PR with UI. 

> I think the "Thru installation" option is a bit too confusing. I would be tempted to take that option out and use the "Specific version" option set to the installation version. Maybe just put in the description text: "Defaults to [version]."

I missed this comment until after posting this version of PR. Yes, I think this will work and reduce the options by one. I'll work on this. 

I have hardcoded the PR to believe that the following versions had css updates. - 1.23.0, 1.28.4 ,1.30.0 ,1.31.5 (these are fake versions and will be removed later). 
I have NOT included any css changes using these selectors. But, this makes it possible for people to test different settings the UI and see selectors being applied and making sure that they are the correct selectors. 

I appreciate code comments at anytime. But, I am still reviewing some of the changes to template.php and theme-settings.php to look for silly mistakes. I would specifically appreciate feedback on naming things, I'm not very familiar with naming best practices in core. 

Is this too long `css_update_version_preference`

**Comment by quicksketch** [comment]
> Is this too long `css_update_version_preference`

No need to say "preference" when this is already "config". Everything is a preference/config/setting. I would shorten to just `css_update_version`.

**Comment by stpaultim** [comment]
@herbdool and @quicksketch 

It's time to look closer at my work. I think I've incorporated most of the suggestions that have been offered. The biggest change is that I'm now doing two things. 

1) Tracking the initial_install version in the system.core.json, which should never change. 
2) Then creating a css_update_version in basis.settings.json that is initially set to be the same as the initial_install version. But, this setting can change. 

In the UI, someone can change the css_update_version to any specific version that had updates or to accept all updates all the time. But, once they change this setting, they can not return it to the install version - they can only select from versions that actually had changes. This makes total sense to me as there is no reason to return this setting to the original default. 

Here are screenshots of UI:

This "Export Options" section is closed by default. Once opened, neither option is selected but behind the scenes the css_updates option is tied to the version of your site at install. If you never find and/or open these settings, then you will always be restricted to any css updates that were available at the time you installed your site. 

![image](https://github.com/user-attachments/assets/eb89215a-abb1-4613-ac85-d23c2df9f00f)

You can then select either "All updates" or "Update thru a specific version." Once you have one of these selected, you must always be using one or the other of these two options. 

![image](https://github.com/user-attachments/assets/5614d578-bede-4c52-9367-3822fec0b3ba)

![image](https://github.com/user-attachments/assets/8c8084ce-d821-47ce-830a-df9272bf7f3f)

**Comment by stpaultim** [comment]
You can test this by playing with the settings and checking the body classes using the inspector. I've hard-coded 4 (fake) versions that had updates, to allow easy testing of this in the sandbox. We'll remove those before merge.

![image](https://github.com/user-attachments/assets/ab5398ee-ce51-4d7f-b95d-70623e4df7ad)

**Comment by quicksketch** [comment]
I like the separation of `install_version` (in System) from `css_update_version` (in Basis). And the settings form in Basis exists where I expect it to be. Putting it in a collapsed fieldset I think is good too. I'm not so sure about calling it "Expert Options" but I'll let others express their preference for terminology there.

> This "Export Options" section is closed by default. Once opened, neither option is selected but behind the scenes the css_updates option is tied to the version of your site at install. If you never find and/or open these settings, then you will always be restricted to any css updates that were available at the time you installed your site.

For me, any time a radio button does not have a default, it makes me apprehensive. "Which one is currently active?" To me it looks like a bug. I would prefer we create a `basis.settings.json` file and set a default to include all updates by default, or set the most current version by default; and we increment that every time we add another update version. No need for any magic calculation.

Also I like the "All updates" option but we lost the "No changes" option. If we want to keep the form short, we can add `- None -` to the version select list, or restore it as a radio button. Either is fine.

**Comment by stpaultim** [comment]
Regarding "Expert Options"

I'm open to suggestions for the title of this collapsed field set. This was just one option thrown out there. Here are a few other ideas:

1) Advanced Options
2) Supplemental CSS
3) CSS Updates

I'd love to hear from others or accept new ideas....

----

Regarding UI options: How about this?

![image](https://github.com/user-attachments/assets/a13db3a1-24fb-44b7-a8bc-6ac22895028c)

I don't think there is any reason to provide an option for rejecting changes prior to installation. This feels like the equivalent of "Not only do I want no updates, but I DO want outdated CSS bugs that were fixed before my installation." This perspective doesn’t seem useful in practice.

HOWEVER, if "No updates" means "No updates after installation," then it makes sense and in my opinion, is the correct default choice.

----

What should the DEFAULT be?

I feel pretty strongly that the default should be to not allow changes after installation, without consent of the site owner. Much of the reason for this PR was that we didn't want to surprise site owners with css changes that inadvertantly effect their site after they do a routine upgrade. To ensure backward compatibility of future CSS changes, I believe that we should only implement them on new sites OR sites that have asked for them.

Working on some changes to the PR that incorporate these suggestions.

**Comment by argiepiano** [comment]
Tiny detail: I believe the correct verb is "affect", as in "they may affect existing sites",  not "effect".  To affect = to have an influence on someone or something, or to cause a change in someone or something. To effect = to achieve something. https://dictionary.cambridge.org/us/grammar/british-grammar/affect-or-effect

**Comment by quicksketch** [comment]
> I don't think there is any reason to provide an option for rejecting changes prior to installation. This feels like the equivalent of "Not only do I want no updates, but I DO want outdated CSS bugs despite the fact that they were fixed before my installation." This perspective doesn’t seem useful in practice.

Considering Basis is made to be subthemed, "No updates" could a useful option for that use-case. Sub-themes may have already fixed the issues present in Basis and don't want any updates ever, past or future. I know some themers despise excessive CSS selectors, and using "No updates" would make it so they could avoid needing any `.update-*` selectors. My intention with suggesting "No updates" is that it would not add any selectors.

> I'm open to suggestions for the title of this collapsed field set.

I like just calling it "CSS Updates" and being consistent. Inside the fieldset are two separate fields related to updates, so seems like a valid title for the group.

> What should the DEFAULT be?

My opinion: It should be "Specific Version" and select the latest available version, i.e. `1.30.0`. Which has the same as effect as "Accept no updates after installation".

My ideal situation would be to have it be "All updates" by default, but if it were sub-themed, then lock to a specific version, but I think that would be tricky to implement and potentially confusing.

**Comment by stpaultim** [comment]
> My ideal situation would be to have it be "All updates" by default, but if it were sub-themed, then lock to a specific version, but I think that would be tricky to implement and potentially confusing.

@quicksketch - You often refer to this as ONLY a problem for sub-theming. I disagree, in fact it's not the use case that concerns me the most. I'm more concerned about the people that might make adjustments in their content to account for weird issues in Basis. 

Example: I set up a simple Backdrop site. I'm not interested in creating a subtheme and my needs are pretty basic. I place a block in the footer with my address in it, but because of a glitch (hypothetical) in Basis, there is not sufficient space above my address. Rather than fixing this with CSS, I just add a blank line above my address. 

Later on, @stpaultim notices this weird glitch in the footer for Basis and fixes it for future users. But, when I download the update, suddenly there is too much space above my address (the FIX plus the space I added in my content to compensate) in the footer for reasons that I don't understand at all - and I may have forgotten about that original space that I added and I'm completely unaware of this "CSS Updates Setting" in Basis.

**Comment by stpaultim** [comment]
@quicksketch 

Granted, "No Updates" might be the preferred option for sub-themers who don't want the supplemental selectors in their markup. I'm fine making that an option. That takes us back to these three options:

No Updates
All Updates
Updates Through Specific Version 

This could work, except I see this problem. In this scenario, I think that the default must be "Updates Through Specific Version", with the most specific version set to be the most recent version at time of installation. 

But, imagine that I'm person like the person I described in the previous comment. I've set up my site and made some hacky adjustments to my site to account for glitches in Basis. It's two years later and numerous "fixes" have been applied to Basis, but I'm not getting them. 

I stumble across this "All Updates" option and think that sounds great. I try it out and severall things seem to break in my theme. I want to revert back to where I started, but I don't remeber which version of updates I was accepting before. 

In my current version, would return to the "no changes" - which takes me back to the default at the time of installation. 

Anyway, we could have 4 options to cover all of these:

1) No supplemental selectors in CSS
2) No changes since installation - (But do includes changes before that)
3) All changes
4) Updates through specific version

I'm going to push my current PR which looks like the following, then wait for more input from other folks.

![image](https://github.com/user-attachments/assets/f973d554-9254-4a66-9d9d-cb290b84c2ea)

**Comment by quicksketch** [comment]
> I'm more concerned about the people that might make adjustments in their content (or other weird configuration hacks) to account for issues in Basis.

Thanks, your explanation also reminded me that there's CSS injector and inline styles if you're using Full HTML. So there are other ways users could get CSS into their site besides a sub-theme too.

> But, this seems like it might be more choices than necessary. Or is it?

We could stash the `- None -` option into the Version select list if we wanted to have it but make it less accessible? Or if `1.5.0` were in the list ([when Basis was introduced](https://github.com/backdrop/backdrop-issues/issues/1361)), that would be the same as none.

I still find the "No updates" option confusing because it _does_ include the updates up to the install version. Could we rename the label to "Default"? Internally referring to it as "no_updates" is fine because it's possible that the "default" behavior may change.

Here's what I would recommend for the labels and descriptions:

![image](https://github.com/user-attachments/assets/4293e202-c00d-446c-b1e7-cf3f989b1ef5)

**Comment by stpaultim** [comment]
One thing that would be really helpful is additional feedback on which options should be available. Which options would you want available and/or which options do you think others would want?

Suggestions for how to describe these options also welcome.

I think @quicksketch has provided some good suggestions. Next version of PR will probably likely look like this (barring additional suggestions):

![image](https://github.com/user-attachments/assets/949ee28c-a2c0-4796-aee7-0c8418b8d0be)

NOTE: We may yet change 1.5.0 to `none`.

**Comment by stpaultim** [comment]
Current version of UI:

![image](https://github.com/user-attachments/assets/e994f0c0-1009-4b14-bda1-4be4c6fc0aa9)

![image](https://github.com/user-attachments/assets/32232430-7be9-4f45-82f1-16c2dfa4cb4b)

**Comment by quicksketch** [comment]
It's looking really great @stpaultim. I like `No updates` more than `1.5.0`, that was sort of a silly suggestion on my part. I left a few more comments in the PR: https://github.com/backdrop/backdrop/pull/4926#pullrequestreview-2507591390

**Comment by jenlampton** [comment]
I'm feeling a lot of scope creep here. 

I would like to propose that we postpone adding any UI at all until after we have some real-world experience with CSS changes in core. It seems like we're doing a lot of guessing about what is going to be needed, and creating a lot of potential confusion in the process. 

The only thing we have on the Basis settings page (without color module) is a setting that a new user won't need on a fresh install - and likely won't understand. What percentage of Backdrop sites are going to benefit from having these options?  Less than 5%? Less than 1%? 

The two use-cases in the original scope of this issue were 1) new installs (that get all fixes up to that point)  and 2) existing sites, which get none of them. There was no need for a UI with that much more limited scope. 

Can these other options be built and improved in contrib before going straight to core?

**Comment by quicksketch** [comment]
> What percentage of Backdrop sites are going to benefit from having these options? Less than 5%? Less than 1%?

I would think 100% of sites running Basis will not want messages to look like this on the homepage (https://github.com/backdrop/backdrop-issues/issues/5769):

![image](https://github.com/user-attachments/assets/431952d6-48cc-4eae-acce-812cdea3122a)

For all existing sites running Basis unmodified, it would be nice if there was a way to decide to opt-in to the messages being fixed without diving into config files or preprocess functions. We'll probably feel similar about other future fixes that might also have a dramatic improvement to the appearance of Basis. I think a lot of sites will want to use the improvements as they are released.

**Comment by jenlampton** [comment]
> I would think 100% of sites running Basis will not want messages to look like this on the homepage (https://github.com/backdrop/backdrop-issues/issues/5769):

And 100% of fresh installs will be fixed -- what percent of existing live Backdrop sites do you think still have that problem? Ones that still have both the default theme + default home page layout, with cards in that location? 

My guess is that this is less than 1% of existing sites.

> For all existing sites running Basis unmodified,

My guess is less than 5% of existing sites.

**Comment by jenlampton** [comment]
> My ideal situation would be to have it be "All updates" by default

This option makes me uneasy because though I see the use case (someone running unmodified Basis without a subtheme) the chances that someone *else* selects this option (should we provide one) would be high, and there is also a very high chance of breaking those sites by accident in the future. 

> Would we ever be including css changes in bugfix releases 

The answer to this should be no (but see below for longer explanation)

> But I don't see any reason inherently to restrict it to minor releases.

This has already been discussed in various other issues. Every time we introduce CSS changes in core, contrib - like the Tatsu theme for example - may also need to have an update. I think it's important that people can count on that not being as frequent as every release, or even in one unscheduled bug fix release. Every 4 months still feels like could be too often, which is why we were talking about doing it only once a year. 

> Once we start tracking install version (accurately) on new installs, we can add classes for any specific version.

It was also mentioned that some front-enders are worried about having too many classes. By limiting the rate of change we will also be limiting the number of classes added to the body tag, which seemed to make everyone feel al little better :)

> One thing that would be really helpful is additional feedback on which options should be available. Which options would you want available and/or which options do you think others would want?

Yes, exactly! So maybe we don't rush into adding the UI? :)

What about a compromise - we leave the config values as they are right now so that people can change them if they want - but hold off on the new UI for a little while. We could even move it into it's own PR so that we don't loose any of the progress that's already been made, and we have it ready to iterate on and improve on?

**Comment by stpaultim** [comment]
In terms of scope creep, I'm not personally invested in adding a UI. I put it into the PR because I was asked and because I'm not necessarily opposed to it and I am frequently disappointed in new additions to core that don't have a UI. A few random thoughts.

1) I agree that much of the thinking behind this PR was about making changes to Basis moving forward, not necessarily making it possible for old sites to pull in changes. So I personally did not expect it to have a UI to start with. But, as I said, I'm a sucker for a UI, so I'm easily convinced and potentially wrong. 

2) I also think it's true, that only a percentage of existing sites are using Basis and only a percentage of those will not have already fixed problems that they didn't like in Basis by default. I personally don't expect too many sites to change this setting. Again, I'm thinking that most changes we will make in the future are for future sites, not old sites. 

3) I suppose that early drafts of this PR had the words "Expert" and "Advanced" on the label for the setting, because I am a bit concerned that some site owners are going to be confused by this setting and potentially use it incorrectly. BUT, I also think we have done a fairly good job of adding help text to minimize that confusion and/or risk and I strongly advocated for a "Default" setting to make it easy for someone to get back to the originally intended setting, if they try something else.

4) I was a bit surprised that we added a UI to this PR, given how often we add features like this without a UI and leave that to either contrib or future releases. But, I am a sucker for a UI, so while surprised, I was not disappointed by the UI. 

5) If we decide against a UI, I will absolutely start working on the contrib module to provide a UI immediately. Usually, I don't feel confident in my ability to create UI's for features like this one. But, in this case, I wrote the possible core UI, so I'm pretty comfortable with the idea of creating the contrib module. I doubt many people will find the contrib module, but we may be surprised and I will feel better knowing it's available.

**Comment by quicksketch** [comment]
The PR looks really good to me. I realized we were missing an update hook to set the defaults for `basis.settings.json`. I pushed a commit that adds a separate update function for Basis in `system.install`.

**Comment by quicksketch** [comment]
@stpaultim Could you try this suggestion I left in the PR? https://github.com/backdrop/backdrop/pull/4926/files#r1890955407

I think we need to make a small change to make it so that sub-themes like Tatsu can also control if CSS updates are applied by Basis.

**Comment by herbdool** [comment]
What happens if Tatsu has an update and wants to set it to a different version to use when someone upgrades Tatsu from contrib? Can it run an update hook?

**Comment by quicksketch** [comment]
> What happens if Tatsu has an update and wants to set it to a different version to use when someone upgrades Tatsu from contrib? Can it run an update hook?

Themes don't have `.install` files or update hooks, so that's not an option. In testing this though I found that if a sub-theme does not provide a `mytheme.settings.json` config file, `theme_get_setting()` looks in the `.info` file instead for a default. So a subtheme can control the Basis behavior by doing the following in the `.info` file:

```
settings[css_update] = custom
settings[css_update_version] = 1.30
```

This value takes precedence over an explicitly set value in the active config `basis.settings.json` if present.

So all-in-all, this seems to work the way I would like it to. When using `theme_get_setting()` it will work in the following order:

- The sub-theme's config file (if any).
- The sub-theme's .info file setting (if any).
- Basis's `basis.settings.json` (guaranteed to have a value now)
- `basis.info` (which doesn't have a value but could)

If a sub-theme wanted to change its value, it could:

- Update the .info file (if it does not have the same value in its config file).
- Provide a UI to change the config value.

So I think `theme_get_setting()` is the right solution here [as suggested in the PR](https://github.com/backdrop/backdrop/pull/4926/files#r1890955407).

p.s. Tatsu is not actually a sub-theme of Basis, it's a copy, so it's not a good theme for testing this. [Thesis](https://backdropcms.org/project/thesis) is a better theme for testing sub-theme functionality.

**Comment by stpaultim** [comment]
It sounds as if you no longer need me to test that suggestion. @quicksketch

**Comment by olafgrabienski** [comment]
I don't have strong feeling about that question if the settings should be avaiable in the user interface. That said, I've had a look at the PR just to see if the settings are still available. They are, and I noticed that the description text of the "Default" option has been removed. Was this by purpose?

**Comment by quicksketch** [comment]
We discussed this for nearly 30 minutes today [during the weekly dev meeting](https://www.youtube.com/live/FepIGE5EoJs?si=jlXxwxfIY1ihUJcS).

The outcome was that we're going to update the PR to remove the UI, but the UI is not ruled out for 1.30.0. Merging the underlying css update system will allow us to proceed with other CSS fixes in Basis, and we'll open a follow-up for the UI.

**Comment by quicksketch** [comment]
@stpaultim I pushed more commits that address some review feedback and today's meeting:

- Changed description text use of "effect" to "affect" per @kswan.
- Fixed the help text missing on the "Default" radio button.
- I renamed the `css_update` config keys to simply `install` (the default), `all`, and `version`, which were easier for me to remember. Added documentation in template.php that these are the available values.
- Added test coverage. I left some `@todo` items in the tests. These pass now _but will intentionally break when we add another CSS update_. That is on purpose so we can expand the test coverage once we have multiple CSS update versions available.
- Removed the extra update versions we were using for manual testing.

I think all that is left is to remove the UI. Let's confirm everything is still working as expected with it available, then remove it as the compromise to move this forward.

**Comment by stpaultim** [comment]
I should be able to update the PR later tonight. I'm working late as usual, but I have a client emergency to deal with first.

**Comment by stpaultim** [comment]
@quicksketch Created a NEW PR without a UI - https://github.com/backdrop/backdrop/pull/4948

Be sure to check the install hook for Basis. I removed what I think was a now unnecessary IF statement. 
https://github.com/backdrop/backdrop/pull/4948/commits/251325e582c24848ec88051f0ff27618b3a0e782

I did not touch the tests, but if you have tests for the UI. I assume they should be removed.

**Comment by stpaultim** [comment]
Also, are we going to remove the CSS example from this PR and deal with the independantly?

**Comment by quicksketch** [comment]
@stpaultim Nice. Could you restore the `css_update` config value? Although it was used in the UI, it was also the only way to have Basis use `"all"` updates.

> I removed what I think was a now unnecessary IF statement https://github.com/backdrop/backdrop/commit/251325e582c24848ec88051f0ff27618b3a0e782

Could you restore this as well? The if statement was never strictly necessary but it is a nice safety check to prevent overwriting a value if one has already been set.

**Comment by stpaultim** [comment]
@quicksketch Yes, I should be able to do that shortly.

**Comment by stpaultim** [comment]
@quicksketch - See: https://github.com/backdrop/backdrop/pull/4948/commits/4aed4a3f6c8bb6432cbc6bdfadec85e658189d6f

**Comment by quicksketch** [comment]
@stpaultim Can you put back the `css_update` `theme_get_setting()` checking in template.php as well? Right now the "all" value doesn't actually work.

**Comment by stpaultim** [comment]
@quicksketch Sure I can put it back, but I'm a bit confused about this "all" setting. Doesn't removing the UI remove the option for "all"? Or is the idea that a site admin could manually change the value in config?

No strong opinions on this, just trying to understand what we're going for.

I did not rewatch the video of dev meeting before I arrived (maybe I should have), so this might have been discussed before I got there.

**Comment by quicksketch** [comment]
> Doesn't removing the UI remove the option for "all"? Or is the idea that a site admin could manually change the value in config?

Yes the option is removed from the UI, but leaving it in place lets an admin change that value to "all", or let's a sub-theme (like a custom sub-theme) set the value to "all" through the theme .info file.

The test coverage checks for both config values (which is why they're now failing). So in short, my expectation is that everything stays identical to how it was with the UI, the only thing removed are the additions to `theme-settings.php`.

**Comment by stpaultim** [comment]
@quicksketch OK, works for me. I just didn't understand it that way, but I did consider it when making the changes. Decided to go with the complete removal, but was not confident in that decision. Fixing now.

**Comment by stpaultim** [comment]
@quicksketch I did a side by side comparison of the previous PR and the new one, the only changes I can see are a comlete removal of the theme-settings.php code and the change of `install-version` to `install` in the Basis update hook (as requested). 

I think it's back to what you intended. 

Just need to decide if we want to leave the CSS change in this PR (I'm fine with it) and/or if there are any tests that should be removed, since the UI is gone.

**Comment by quicksketch** [comment]
> Just need to decide if we want to leave the CSS change in this PR (I'm fine with it) 

This issue is so long already, I think we should remove the CSS and file a PR specifically against https://github.com/backdrop/backdrop-issues/issues/2260, which seems to be the related issue.

> if there are any tests that should be removed, since the UI is gone.

Nope, tests were written with the removal of the UI in mind. It just tests the APIs. Latest PR is passing all the new tests again.

**Comment by stpaultim** [comment]
> This issue is so long already, I think we should remove the CSS and file a PR specifically against https://github.com/backdrop/backdrop-issues/issues/2260, which seems to be the related issue.

Yes, I agree. I will come back to that later.

For now, I have removed the CSS changes to Basis from this PR. @quicksketch Back to you.

**Comment by quicksketch** [comment]
Looks perfect.

Effectively, the only thing this does at this point is add `update-1-30` to the `<body class>` attribute. :sweat_smile: 

It adds install_version into Telemetry too.
![image](https://github.com/user-attachments/assets/c8fde4da-b363-4262-a7ac-4acbfc080c7b)

But that's really about it.

With test coverage checking for the existence of the class and that different options remove or display the additional class, I think this is good to go.

@olafgrabienski @jenlampton @herbdool (et al) would one of you be willing to give this the final review?

**Comment by herbdool** [comment]
I reviewed https://github.com/backdrop/backdrop/pull/4948 which I assume is the right one. It looks good to go now.

**Comment by stpaultim** [comment]
@herbdool Yes, this is the correct PR: https://github.com/backdrop/backdrop/pull/4948

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/4948 into 1.x for 1.30.0!!

Let the CSS fixes begin! I think there's a useful meta issue already at #4094.

Thank you everyone!

https://github.com/backdrop/backdrop/commit/66b7f892f0bf050ddcb92fc9333454cfd8867b9c by @stpaultim, @jenlampton, @quicksketch, @herbdool, @olafgrabienski, @klonos, @BWPanda, @argiepiano, and @kswan.
